design-proposals/kubebmc.md

```diff
@@ -0,0 +1,156 @@
+- As a tester, I need to quickly spin up and destroy OpenStack clusters in a virtual environment for testing purposes.
+
+## Repos
+
+No existing repos under the KubeVirt organization will be impacted. KubeBMC augments the KubeVirt ecosystem as an add-on.
+
+# Design
+
+KubeBMC will act as a bridge between KubeVirt VMs and the BMC functionalities, using Kubernetes services to expose these capabilities. KubeBMC consists of two components:
+
+- **kubebmc-controller**: A typical Kubernetes controller reconciles the KubeBMC, VirtualMachine, and Service objects
+- **kbmc**: A BMC simulator for serving IPMI and Redfish endpoints and translating the incoming requests to native Kubernetes API calls
+
+Each VirtualMachine will have its own KubeBMC object created by `kubebmc-controller` during the VM creation. The controller will further process the KubeBMC object to complete the BMC scaffold. This includes a BMC emulator Pod, i.e., `kbmc`, and its Service object working together to provide the BMC functionalities to various clients over the network.
+
+Inside the BMC emulator, the IPMI/Redfish requests are translated to native Kubernetes API calls and sent to the API server. It's essentially a combination of IPMI/HTTP server and Kubernetes client.
+
+The following graph illustrates the entire KubeBMC workflow:
+
+```mermaid
+flowchart LR
+    controller["kubebmc-controller"]
+    cr["kubebmc CR"]
+    kbmc-pod["kbmc Pod"]
+    kbmc-svc["kbmc Service"]
+    controller-.->|watches|cr
+    cr-.->|owns|kbmc-svc
+    cr-.->|owns|kbmc-pod
+    client--->|IPMI|kbmc-svc
+    kbmc-svc-->kbmc-pod
+    kbmc-pod-->|HTTP|apiserver
+    apiserver-->|modifies|vm
+    vm-->|creates|vmi
+```
+
+It's also possible for VMs being created to opt out of such a functionality. This could be done by annotating the VirtualMachine object with a specific key-value pair.
+
+## API Examples
+
+The proposed KubeBMC CRD looks like the following:
+
+```yaml
+kind: KubeBMC
+metadata:
+  name: openshift-cluster-node-0
+spec:
+  # A Secret object reference for the IPMI/Redfish authentication.
+  # It should contain a pair of username and password.
+  authSecretRef:
+    namespace: default
+    name: bmc-cred
+  # The associated VM
+  virtualMachineRef:
+    namespace: default
+    name: openshift-cluster-node-0
+```
+
+The condition of the created resources (Pod and Service) will be reflected under the `.status` section:
+
+```yaml
+status:
+  serviceIP: 10.53.23.74
+  conditions:
+  - lastUpdateTime: "2024-03-01T08:48:23Z"
+    status: "True"
+    type: BMCEmulatorReady
+  - lastUpdateTime: "2024-03-01T08:48:23Z"
+    type: EndpointReady
+    status: "True"
+```
+
+## Scalability
+
+KubeBMC is designed to scale within Kubernetes environments, and the number of emulated BMC Pods and Services should increase as the number of VM increases. Hence, the overhead each Pod introduced to the entire system should also be considered.
```

Comment:
I wonder if we could make this a sidecar to the VM?

Reply:
Yes. Then the emulated BMC could operate the VM by directly communicating with `virtqemud` via the Unix socket. No communication with `kube-apiserver` is needed anymore. However, the entire architecture will be different from the current design of KubeBMC and more coupled with KubeVirt itself because users might need to determine if they want a virtualized BMC sidecar container created alongside the VM at the beginning.

That's my initial thoughts about it. Maybe it's not that complicated. Please correct me if there's anything wrong. Thank you.

Reply:
Thinking about it again, this might not be so easy. Because a sidecar can only be attached to a VMI (a running VM), while the BMC also needs to be available when the VM is stopped. (how do you start it otherwise?)

Also, you are right, it is more decoupled right now and should probably stay like that to keep the VM "lean".

Reply:
Nice catch! If we go with the sidecar design, powering on an already shutdown VM via BMC wouldn't be possible. The BMC serves as an out-of-band manner for users/programs to manage the host itself; that's how it was initially targeted.